### DISCUSSION TITLE
Securing a chat message table?

### DISCUSSION THREAD

**Discussion by GaryAustin1** [supabase/supabase#3500] [Questions]
So I have a table with messages in many  different chat rooms.  The messages are indexed on the chat room lower case name which is equivalent to a unique password.  Any user should only be able to read messages they know the chat room name of.  

I can't figure a way to use RLS to look at the where filter coming in with the select and make sure it has the chat room key column and limit rows to those.  With out this, an authorized could technically do a select without a filter and get all rows.  

Two other approaches I'm looking at is locking down the message table and only using functions to access it where you provide the chat room key, or in the RLS policy use a function to look up the room keys you have access to in another table that exists for rooms you have joined.

Any one deal with something similar?

**Comment by laem** [comment]
Hi Gary, I'm having the exact same issue. 
I've found this answered question that might help us : https://github.com/supabase/supabase/discussions/5033

edit : just understood that you are the one giving the answer. Did it resolve your usecase ? 

The problem for me with the function approach is : can you subscribe to "insert" and "update" events ? 

edit : I guess it will be resolved by hooks calling functions, not available yet 

![image](https://user-images.githubusercontent.com/1177762/150319852-3a9b14ed-8f58-4123-93bf-929c61230ee1.png)

**Reply by laem** [reply]
Here's the function I had to write. 
Supabase's function editor does not allow to return a set of records, hence you need to write it yourself in the SQL editor. 

```
create or replace function get_answers(value text) 
returns setof réponses
language sql
security definer
as $$
  select * from réponses where sondage = value;
$$;

```

**Reply by GaryAustin1** [reply]
@laem Thanks for the reply.  In this question I was looking for ways people were handle it as there are several and each has tradeoffs.
I thought about using your type of function but ended up going with a "join" table that has all the chat rooms a user has joined.  Then I have a function that returns all your chat rooms by user id, so in RLS I can do  using ((room_key)::text = ANY ((my_rooms())::text[]));
This approach was better for me than just a "password" match to a chatroom because I use the join table to also show who is in the room so you can notice if some one got in that you did not want there.

**Reply by laem** [reply]
Thanks, it looks quite complicated for my use case which is a bit simpler : I don't want to link the multiple "rooms" that a person has joined in my database. It's a survey use case, I've described it here more in detail : https://github.com/supabase/supabase/discussions/5073
